internal/primitives/runtime/interfaces.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package runtime
+
+import (
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hashing"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"golang.org/x/exp/constraints"
+)
+
+// Number is the header number type
+type Number interface {
+	~uint | ~uint32 | ~uint64
+}
+
+// Hash type
+type Hash interface {
+	constraints.Ordered
+	// Bytes returns a byte slice representation of Hash
+	Bytes() []byte
+	// String returns a unique string representation of the hash
+	String() string
+}
+
+// Hasher is an interface around hashing
+type Hasher[H Hash] interface {
+	// Produce the hash of some byte-slice.
+	Hash(s []byte) H
+
+	// Produce the hash of some codec-encodable value.
+	HashOf(s any) H
```

Comment:
Could these be combined?

Reply:
`HashOf` is SCALE encoding `s` before hashing.  `Hash` is just hashing any `[]byte`.

Reply:
I'm open to suggestions.  It makes sense to me.  `hasher.HashOf(myThing)` means give me a hash of this type.  `SCALEEncodedHashOf`? 😂

Reply:
then lests call it EncodeAndHash, this name exactly says what is going to happen inside

Reply:
I don't see how `Hasher` violates the single responsibility pattern, given that it will hash the SCALE encoding of a type.  I'll revise the name to `HashEncoded`.  Capitalized `Of` or `And` doesn't smell right to me.